docs/posts/1brc/index.qmd

```diff
@@ -0,0 +1,336 @@
+
+```{python}
+t = ibis.read_csv("1brc/data/measurements.txt", **kwargs)
+t.limit(3)
+```
+
+Then let's confirm it's **a billion** rows:
+
+```{python}
+f"{t.count().to_pandas():,}"
+```
+
+Finally, we'll compute the min, mean, and max temperature per weather station:
+
+```{python}
+res = run_challenge(t)
+res
+```
+
+## Polars
+
+First let's set the backend to Polars and the `kwargs` dictionary:
+
+```{python}
+ibis.set_backend("polars") # <1>
+kwargs = polars_kwargs
+```
+
+1. Set Polars as the default backend used
+
+Next, we'll read in the data and take a look at the table:
+
+```{python}
+t = ibis.read_csv("1brc/data/measurements.txt", **kwargs)
+t.limit(3)
+```
+
+Then let's confirm it's **a billion** rows:
+
+```{python}
+f"{t.count().to_pandas():,}"
+```
+
+Finally, we'll compute the min, mean, and max temperature per weather station:
+
+```{python}
+res = run_challenge(t)
+res
+```
+
+## DataFusion
+
+First let's set the backend to DataFusion and the `kwargs` dictionary:
+
+```{python}
+ibis.set_backend("datafusion") # <1>
+kwargs = datafusion_kwargs
+```
+
+1. Set DataFusion as the default backend used
+
+Next, we'll read in the data and take a look at the table:
+
+```{python}
+t = ibis.read_csv("1brc/data/measurements.txt", **kwargs)
+t.limit(3)
+```
+
+Then let's confirm it's **a billion** rows:
+
+```{python}
+f"{t.count().to_pandas():,}"
+```
+
+Finally, we'll compute the min, mean, and max temperature per weather station:
+
+```{python}
+res = run_challenge(t)
+res
+```
+
+:::
+
+## Bonus: more billion row data generation
+
+While we're here, I'll share the code I've used in the past to generate a
+billion rows of random data:
+
+```{.python}
+import ibis
+
+con = ibis.connect("duckdb://data.ddb")
+
+ROWS = 1_000_000_000
+
+sql_str = ""
+sql_str += "select\n"
+for c in list(map(chr, range(ord("a"), ord("z") + 1))):
+    sql_str += f"  random() as {c},\n"
+sql_str += f"from generate_series(1, {ROWS})"
+
+t = con.sql(sql_str)
+con.create_table("billion", t, overwrite=True)
+```
+
+Nowadays I'd convert that to an Ibis expression:
+
+:::{.callout-note}
+This is a slightly different result with a monotonic index column, but I prefer
+it anyway. You could drop that column or adjust the expression.
+:::
+
+```{.python}
+import ibis
+
+con = ibis.connect("duckdb://data.ddb")
+
+ROWS = 1_000_000_000
+
+t = (
+    ibis.range(ROWS)
+    .unnest()
+    .name("index")
+    .as_table()
+    .mutate(**{c: ibis.random() for c in list(map(chr, range(ord("a"), ord("z") + 1)))})
+)
+con.create_table("billion", t, overwrite=True)
+```
+
+But if you do need to construct a programmatic SQL string, it's cool that you
+can!
+
+## Conclusion
+
+While the one billion row challenge isn't a great benchmark, it's a fun way to
+demonstrate how Ibis provides a single Python dataframe API to take the billion
+row challenge with DuckDB, Polars, and DataFusion. Feel free to try it out with
+other backends!
+
+Happy coding!
```

Comment:
This should fix the linting issue https://github.com/ibis-project/ibis/actions/runs/7572830475/job/20623574885?pr=8004
```suggestion
Happy coding!

```

Reply:
yep just have to re-run to get the execution output, about to push

Reply:
also had to regenerate the data because I accidentally deleted it